### PR TITLE
nixos/simplesamlphp: init

### PR DESCRIPTION
**Pull request by thenhnn**
## Description of changes

Packaging of [SimpleSAMLphp](https://simplesamlphp.org/), an application written in native PHP that deals with authentication.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by balsoft — APPROVED** [review]

File: pkgs/pkgs-lib/formats.nix
Comment (infinisil) [inline pkgs/pkgs-lib/formats.nix:280]: Shouldn't there be a closing tag?

File: pkgs/pkgs-lib/formats.nix
Comment (infinisil) [inline pkgs/pkgs-lib/formats.nix:352]: Taking a step back, can't we just encode to a JSON file and generate some static PHP to decode that? I don't write PHP, but something like

```nix
text = ''
  <?php 
  $json_data = json_decode(file_get_contents('${json.generate "${name}-json" value}'),true); 
  ?>
''
```

This way we can also reuse `pkgs.format.json`'s `type` and semantics.

**Review by infinisil — CHANGES_REQUESTED** [review]

File: pkgs/pkgs-lib/formats.nix
Comment (thenhnn) [reply pkgs/pkgs-lib/formats.nix:280]: Nope, [see this](https://www.php.net/manual/en/language.basic-syntax.phptags.php#:~:text=If%20a%20file,in%20the%20script.).

File: pkgs/pkgs-lib/formats.nix
Comment (thenhnn) [reply pkgs/pkgs-lib/formats.nix:306]: Okay, it's just a PHP thing that array can contain both key-value pairs and just array items.

I don't know how to name it to not conflict with something else. Do you have any ideas?

File: pkgs/pkgs-lib/formats.nix
Comment (thenhnn) [reply pkgs/pkgs-lib/formats.nix:352]: We can't represent array that can be both an ordinary list and an key-value map with JSON, that's why I made this code.

File: pkgs/pkgs-lib/formats.nix
Comment (thenhnn) [reply pkgs/pkgs-lib/formats.nix:347]: Thanks for the clarification.

File: pkgs/pkgs-lib/formats.nix
Comment (thenhnn) [reply pkgs/pkgs-lib/formats.nix:319]: PHP array can be both ordinary list **and** key-value map, as [described in PHP docs](https://www.php.net/manual/en/language.types.array.php). I should probably split this into functions, to make this more readable.

File: pkgs/pkgs-lib/formats.nix
Comment (thenhnn) [reply pkgs/pkgs-lib/formats.nix:278]: I probably can make separate PR, but I'm not sure if I can merge it from separate PR to this and Filesender properly. I will try.

File: pkgs/pkgs-lib/formats.nix
Comment (infinisil) [reply pkgs/pkgs-lib/formats.nix:352]: What's the problem with converting JSON objects to associative arrays and JSON lists to indexed arrays?

Comment (thenhnn) [reply pkgs/pkgs-lib/formats.nix:352]: There's no problem if it's only **assoctiative array** or **indexed array**. But PHP allows array to be both at the same time.

Currently it's looks like this:
```nix
default = {
   "" = [ "exampleauth:UserPass" ];
   "user:password" = {
       uid = [ "user" ];
       cn = [ "user" ];
       mail = [ "user@nixos.org" ];
   };
};
```

File: pkgs/pkgs-lib/formats.nix
Comment (thenhnn) [reply pkgs/pkgs-lib/formats.nix]: I removed it, so we can wait until #179471 is merged to add it back, later. It is possible to n use `format.lib.mkRaw "eval(file_get_contents(\"${mySecretPath}\"))"` as workaround. Should I add it to docs?

File: pkgs/pkgs-lib/formats.nix
Comment (thenhnn) [reply pkgs/pkgs-lib/formats.nix]: I created helper function `format.lib.mkMixedArray` that implements that without empty string attribute hack.

File: pkgs/pkgs-lib/formats.nix
Comment (thenhnn) [reply pkgs/pkgs-lib/formats.nix]: I would like to stay it in one PR if it's not critical. I'm afraid that it will cause a lot of problems with second PR.

**Comment by thenhnn** [comment]
@infinisil Can you check the comments on the review? (I'm just not sure if GitHub publishes notifications about review comments).

**Comment by thenhnn** [comment]
@infinisil hello, can you check out comments on review, please?

File: pkgs/pkgs-lib/formats.nix
Comment (thenhnn) [reply pkgs/pkgs-lib/formats.nix]: Probably no.

**Pull request closed without merging**
